ERROR: Mismatch in directionality for cell port...

## Steps to reproduce the issue

Possibly the same issue as https://github.com/YosysHQ/yosys/issues/2053

My code uses arrays where some lines are inputs and some are outputs.  With recent Yosys builds, some small part of this code is that used to work for a long time is no-longer compiling.  The code still functions in Vivado, fyi.

Boiled down, the following snippet illustrates this.

There is a 9-bit array, which is marked as `inout` because of the mix of signals.  The subordinate module, `worker` assigns the parts of the bundle.

``` verilog

module worker(
        inout  [8:0] data_in,
        inout  [8:0] data_out
    );
    assign data_out[7:0] = data_in[7:0];
    assign data_in[ 8 ] = data_out[ 8 ];
endmodule

module tests_tbx (
        inout [8:0] data
    );
    wire [8:0] data_shadow;
    assign data_shadow[7:0] = 8'H55;
    worker w( data_shadow, data );
endmodule

```
## Expected behavior

This should compile.

## Actual behavior

```
$ make
yosys -q -p 'synth_ice40 -top tests_tbx -json tests_tbx.json' ../projects/tests_tbx/tests_tbx.v
ERROR: Mismatch in directionality for cell port tests_tbx.w.data_in: \data_shadow <= \w.data_in
Makefile:30: recipe for target 'tests_tbx.json' failed
make: *** [tests_tbx.json] Error 1
```

Yep

https://github.com/YosysHQ/yosys/commit/f88bef767263590c94e157d0989afa91db3ccdb0 - fails

```
$ yosys --version
Yosys 0.9+2406 (git sha1 f88bef76, clang 6.0.0-1ubuntu2 -fPIC -Os)
```

```
$ make
yosys -q -p 'synth_ice40 -top tests_tbx -json tests_tbx.json' ../projects/tests_tbx/tests_tbx.v
ERROR: Mismatch in directionality for cell port tests_tbx.w.data_in: \data_shadow <= \w.data_in
Makefile:30: recipe for target 'tests_tbx.json' failed
make: *** [tests_tbx.json] Error 1
```

I'll check that next, Dave.  Meanwhile...

April 24 - https://github.com/YosysHQ/yosys/commit/bbf343589ba2b4508afdfaed17dfd11c3252cf7f - fails
```
$ yosys --version
Yosys 0.9+2406 (git sha1 bbf34358, clang 6.0.0-1ubuntu2 -fPIC -Os)
```

```
$ make
yosys -q -p 'synth_ice40 -top tests_tbx -json tests_tbx.json' ../projects/tests_tbx/tests_tbx.v
ERROR: Mismatch in directionality for cell port tests_tbx.w.data_in: \data_shadow <= \w.data_in
Makefile:30: recipe for target 'tests_tbx.json' failed
make: *** [tests_tbx.json] Error 1
```

I'll check the one prior.

Feb 25 - https://github.com/YosysHQ/yosys/pull/1721/commit/5cae9c6e1628db7223a205d2444ef26e682007a5 - fails

```
$ yosys --version
Yosys 0.9+1706 (git sha1 5cae9c6e, clang 6.0.0-1ubuntu2 -fPIC -Os)
```
```
$ make
yosys -q -p 'synth_ice40 -top tests_tbx -json tests_tbx.json' ../projects/tests_tbx/tests_tbx.v
ERROR: Mismatch in directionality for cell port tests_tbx.w.data_in: \data_shadow <= \w.data_in
Makefile:30: recipe for target 'tests_tbx.json' failed
make: *** [tests_tbx.json] Error 1
```

Uh oh...

Feb 22 - #1715 - https://github.com/YosysHQ/yosys/commit/6edca05793197a846bbfb0329e836c87fa5aabb6 - fails

```
$ yosys --version
Yosys 0.9+1706 (git sha1 6edca057, clang 6.0.0-1ubuntu2 -fPIC -Os)
```
```
$ make
yosys -q -p 'synth_ice40 -top tests_tbx -json tests_tbx.json' ../projects/tests_tbx/tests_tbx.v
ERROR: Mismatch in directionality for cell port tests_tbx.w.data_in: \data_shadow <= \w.data_in
Makefile:30: recipe for target 'tests_tbx.json' failed
make: *** [tests_tbx.json] Error 1
```

OK.  0.9 next.  Meanwhile...

 Jan 30 - https://github.com/YosysHQ/yosys/commit/60876ce183ee5f3980c378e190d996453d59a780 - fails

```
$ yosys --version
Yosys 0.9+1706 (git sha1 6edca057, clang 6.0.0-1ubuntu2 -fPIC -Os)
```

```
$ make
yosys -q -p 'synth_ice40 -top tests_tbx -json tests_tbx.json' ../projects/tests_tbx/tests_tbx.v
ERROR: Mismatch in directionality for cell port tests_tbx.w.data_in: \data_shadow <= \w.data_in
Makefile:30: recipe for target 'tests_tbx.json' failed
make: *** [tests_tbx.json] Error 1
```

Dec 31 - https://github.com/YosysHQ/yosys/commit/c082329af33cd428f53f5afbcb51fab8de545090 - fails

```
$ yosys -V
Yosys 0.9+932 (git sha1 c082329a, clang 6.0.0-1ubuntu2 -fPIC -Os)
```
```
$ make
yosys -q -p 'synth_ice40 -top tests_tbx -json tests_tbx.json' ../projects/tests_tbx/tests_tbx.v
ERROR: Mismatch in directionality for cell port tests_tbx.w.data_in: \data_shadow <= \w.data_in
Makefile:30: recipe for target 'tests_tbx.json' failed
make: *** [tests_tbx.json] Error 1
```

Something weird is going on. The test case fails here, but my actual code does compile.  I suspected it would because I know I have pulled Yosys since Nov 1.  

Nov 1 - https://github.com/YosysHQ/yosys/commit/c4bd318e76240d3e6a95109c19641cdfd86517b8 - fails

```
$ yosys -V
Yosys 0.9+932 (git sha1 c4bd318e, clang 6.0.0-1ubuntu2 -fPIC -Os)
```
```
$ make
yosys -q -p 'synth_ice40 -top tests_tbx -json tests_tbx.json' ../projects/tests_tbx/tests_tbx.v
ERROR: Mismatch in directionality for cell port tests_tbx.w.data_in: \data_shadow <= \w.data_in
Makefile:30: recipe for target 'tests_tbx.json' failed
make: *** [tests_tbx.json] Error 1
```

Here's a tiny test that shows the error between the two builds.  

``` verilog 
module subworker(
        inout  [8:0] pipe_in,
        inout  [8:0] pipe_out
    );
    assign pipe_out[7:0] = pipe_in[7:0];
    assign pipe_in[ 8 ] = pipe_out[ 8 ];
endmodule

module worker (
        inout [8:0] pipe
    );
    wire [8:0] pipe_shadow;
    assign pipe_shadow[7:0] = 8'H55;
    subworker sw( pipe_shadow, pipe );
    assign pipe[8] = 1'B1;
endmodule

module tests_tbx (
        output [7:0] data,
        output      control
    );

    wire [8:0] pipe;

    assign data = pipe[7:0];
    assign control = pipe[ 8 ];

    worker w( pipe );
endmodule
```

Then here's what building with the two branches results in

```
$ yosys --version
Yosys 0.9+2406 (git sha1 467152d7, clang 6.0.0-1ubuntu2 -fPIC -Os)
$ make clean
rm -f tests_tbx.json tests_tbx.asc tests_tbx.rpt tests_tbx.bin pll.v
$ make
yosys -q -p 'synth_ice40 -top tests_tbx -json tests_tbx.json' ../projects/tests_tbx/tests_tbx.v
nextpnr-ice40 --lp8k --freq 48 --top tests_tbx --package cm81 --pcf ../projects/tests_tbx/pins.pcf --json tests_tbx.json --asc tests_tbx.asc
$ # yosys has completed, no error
```

Switching to other toolset

```
$ yosys --version
Yosys 0.9+2406 (git sha1 4ef9ee3c, clang 6.0.0-1ubuntu2 -fPIC -Os)
$ make clean
rm -f tests_tbx.json tests_tbx.asc tests_tbx.rpt tests_tbx.bin pll.v
$ make
yosys -q -p 'synth_ice40 -top tests_tbx -json tests_tbx.json' ../projects/tests_tbx/tests_tbx.v
ERROR: Mismatch in directionality for cell port worker.sw.pipe_in: \pipe_shadow <= \sw.pipe_in
Makefile:30: recipe for target 'tests_tbx.json' failed
make: *** [tests_tbx.json] Error 1
```

--- a/passes/techmap/flatten.cc
+++ b/passes/techmap/flatten.cc
@@ -152,15 +152,14 @@ struct FlattenWorker
 
 		// Attach port connections of the flattened cell
 
-		SigMap tpl_sigmap(tpl);
 		pool<SigBit> tpl_driven;
 		for (auto tpl_cell : tpl->cells())
 			for (auto &tpl_conn : tpl_cell->connections())
 				if (tpl_cell->output(tpl_conn.first))
-					for (auto bit : tpl_sigmap(tpl_conn.second))
+					for (auto bit : tpl_conn.second)
 						tpl_driven.insert(bit);
 		for (auto &tpl_conn : tpl->connections())
-			for (auto bit : tpl_sigmap(tpl_conn.first))
+			for (auto bit : tpl_conn.first)
 				tpl_driven.insert(bit);
 
 		SigMap sigmap(module);
@@ -190,7 +189,7 @@ struct FlattenWorker
 			} else {
 				SigSpec sig_tpl = tpl_wire, sig_mod = port_it.second;
 				for (int i = 0; i < GetSize(sig_tpl) && i < GetSize(sig_mod); i++) {
-					if (tpl_driven.count(tpl_sigmap(sig_tpl[i]))) {
+					if (tpl_driven.count(sig_tpl[i])) {
 						new_conn.first.append(sig_mod[i]);
 						new_conn.second.append(sig_tpl[i]);
 					} else {

--- a/passes/techmap/techmap.cc
+++ b/passes/techmap/techmap.cc
@@ -233,16 +233,14 @@ struct TechmapWorker
 			}
 		}
 
-		SigMap tpl_sigmap(tpl);
 		pool<SigBit> tpl_written_bits;
-
 		for (auto tpl_cell : tpl->cells())
 		for (auto &conn : tpl_cell->connections())
 			if (tpl_cell->output(conn.first))
-				for (auto bit : tpl_sigmap(conn.second))
+				for (auto bit : conn.second)
 					tpl_written_bits.insert(bit);
 		for (auto &conn : tpl->connections())
-			for (auto bit : tpl_sigmap(conn.first))
+			for (auto bit : conn.first)
 				tpl_written_bits.insert(bit);
 
 		SigMap port_signal_map;
@@ -280,7 +278,7 @@ struct TechmapWorker
 				SigSpec sig_tpl = w, sig_tpl_pf = w, sig_mod = it.second;
 				apply_prefix(cell->name, sig_tpl_pf, module);
 				for (int i = 0; i < GetSize(sig_tpl) && i < GetSize(sig_mod); i++) {
-					if (tpl_written_bits.count(tpl_sigmap(sig_tpl[i]))) {
+					if (tpl_written_bits.count(sig_tpl[i])) {
 						c.first.append(sig_mod[i]);
 						c.second.append(sig_tpl_pf[i]);
 					} else {

--- a/tests/techmap/bug2183.ys
+++ b/tests/techmap/bug2183.ys
@@ -1,0 +1,11 @@
+read_verilog <<EOT
+module foo(inout a, b);
+  assign a = b;
+endmodule
+module bar(output c);
+  foo f(c, 1'b0);
+endmodule
+EOT
+
+hierarchy -auto-top
+flatten

